evennia/contrib/wilderness.py

```diff
@@ -0,0 +1,490 @@
+
+"""
+
+from evennia import DefaultRoom, DefaultExit, DefaultScript
+from evennia import create_object, create_script
+from evennia.utils import inherits_from
+
+
+class WildernessRoom(DefaultRoom):
+
+    @property
+    def wilderness(self):
+        """Shortcut property to the wilderness script this room belongs to."""
+        return self.ndb.wildernessscript
+
+    @property
+    def location_name(self):
+        """Shortuct property to the location name."""
+        return self.wilderness.provider.get_location_name(
+            self.ndb.active_location)
+
+    def at_object_receive(self, moved_obj, source_location):
+        """Called after an object has been moved into this object."""
+        if moved_obj.destination and moved_obj.destination == moved_obj.location:
+            # Ignore exits looping back to themselves: those are the regular
+            # n, ne, ... exits.
+            return
+
+        itemlocs = self.wilderness.db.itemlocations
+        if moved_obj in itemlocs:
+            # This object was already in the wilderness. We need to make sure it
+            # goes to the correct room it belongs to.
+            # Otherwise the following issue can come up:
+            # 1) Player 1 and Player 2 share a room
+            # 2) Player 1 disconnects
+            # 3) Player 2 moves around
+            # 4) Player 1 reconnects
+            # Player 1 will end up in player 2's room, which has the wrong
+            # location
+
+            location = itemlocs[moved_obj]
+            # Setting the location to None is important here so that we always
+            # get a "fresh" room
+            moved_obj.location = None
+            self.wilderness.move_obj(moved_obj, location)
+        else:
+            # This object wasn't in the wilderness yet. Let's add it.
+            itemlocs[moved_obj] = self.ndb.active_location
+
+    def at_object_leave(self, moved_obj, target_location):
+        """Called after an object has left this object."""
+        self.wilderness.at_after_object_leave(moved_obj)
+
+    def set_active_location(self, new_location):
+        """Changes this room to show the conect of new_location.
+
+        Args:
+            new_location ((int, int)): location as tuple of (x, y)
+        """
+        # Remove the reference for the old location...
+        rooms = self.wilderness.db.rooms
+        del rooms[self.ndb.active_location]
+        # ...and add it for the new location.
+        self.ndb.active_location = new_location
+        rooms[self.ndb.active_location] = self
+
+        # Every obj inside this room will get its location set to None
+        for item in self.contents:
+            if not item.destination or item.destination != item.location:
+                item.location = None
+        # And every obj matching the new location will get its location set to
+        # this room
+        for item, location in self.wilderness.db.itemlocations.iteritems():
```

Comment:
The performance of this will scale linearly (O(N)) with the number of the objects in the wilderness. Which may be fine; for a larger set one could consider using Tags for the coordinates. Those scale as O(1) due to database indexing and would eventually be considerably more effective. Not something for this PR, but maybe worth to make a comment on for those having bigger performance requirements.